evm/src/cross_table_lookup.rs

```diff
@@ -649,3 +649,123 @@ pub(crate) fn verify_cross_table_lookups_circuit<
     }
     debug_assert!(ctl_zs_openings.iter_mut().all(|iter| iter.next().is_none()));
 }
+
+#[cfg(test)]
+pub(crate) mod testutils {
+    use std::collections::HashMap;
+
+    use plonky2::field::polynomial::PolynomialValues;
+    use plonky2::field::types::Field;
+
+    use crate::all_stark::Table;
+    use crate::cross_table_lookup::{CrossTableLookup, TableWithColumns};
+
+    type MultiSet<F> = HashMap<Vec<F>, Vec<(Table, usize)>>;
+
+    /// Check that the provided traces and cross-table lookups are consistent.
+    pub(crate) fn check_ctls<F: Field>(
+        trace_poly_values: &[Vec<PolynomialValues<F>>],
+        cross_table_lookups: &[CrossTableLookup<F>],
+    ) {
+        for (i, ctl) in cross_table_lookups.iter().enumerate() {
+            check_ctl(trace_poly_values, ctl, i);
+        }
+    }
+
+    fn check_ctl<F: Field>(
+        trace_poly_values: &[Vec<PolynomialValues<F>>],
+        ctl: &CrossTableLookup<F>,
+        ctl_index: usize,
+    ) {
+        let CrossTableLookup {
+            looking_tables,
+            looked_table,
+            default,
+        } = ctl;
+
+        // Maps `m` with `(table, i) in m[row]` iff the `i`-th row of `table` is equal to `row` and
+        // the filter is 1. Without default values, the CTL check holds iff `looking_multiset == looked_multiset`.
+        let mut looking_multiset = MultiSet::<F>::new();
+        let mut looked_multiset = MultiSet::<F>::new();
+
+        for table in looking_tables {
+            process_table(trace_poly_values, table, &mut looking_multiset);
+        }
+        process_table(trace_poly_values, looked_table, &mut looked_multiset);
+
+        let empty = &vec![];
+        // Check that every row in the looking tables appears in the looked table the same number of times
+        // with some special logic for the default row.
+        let mut extra_default_count = default.as_ref().map(|_| 0);
+        for (row, looking_locations) in &looking_multiset {
+            let looked_locations = looked_multiset.get(row).unwrap_or(empty);
+            if let Some(default) = default {
+                if row == default {
+                    *extra_default_count.as_mut().unwrap() +=
+                        looking_locations.len() - looked_locations.len();
+                    continue;
+                }
+            }
+            check_locations(looking_locations, looked_locations, ctl_index, row);
+        }
+        // Check that the number of extra default rows is correct.
```

Comment:
It seems like this could fail if the default row occurs naturally (with a matching row in the looked table). I would suggest adding the table size delta to `looked_multiset[default]` and then proceeding with the normal multiset comparison.

Reply:
Oh I see, I misread the `extra_default_count` calculation.

Seems right then, though I think it'd be slightly easier to read if we calculate it upfront instead of in the loop, like
```rust
        let extra_default_count = default.as_ref().map(|d| {
            let looking_defaults = looking_multiset.get(d).unwrap_or(empty).len();
            let looked_defaults = looked_multiset.get(d).unwrap_or(empty).len();
            looking_defaults - looked_defaults
        });
```

Reply:
Good idea, done!